node/core/approval-voting/src/criteria.rs

```diff
@@ -0,0 +1,339 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Assignment criteria VRF generation and checking.
+
+use polkadot_node_primitives::approval::{
+	self as approval_types, AssignmentCert, AssignmentCertKind, DelayTranche, RelayVRFStory,
+};
+use polkadot_primitives::v1::{CoreIndex, ValidatorIndex, SessionInfo, AssignmentPair};
+use sc_keystore::LocalKeystore;
+use parity_scale_codec::{Encode, Decode};
+use sp_application_crypto::Public;
+
+use merlin::Transcript;
+use byteorder::{ByteOrder, LE};
+use schnorrkel::vrf::VRFInOut;
+
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+
+use super::LOG_TARGET;
+
+/// Details pertaining to our assignment on a block.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct OurAssignment {
+	cert: AssignmentCert,
+	tranche: DelayTranche,
+	validator_index: ValidatorIndex,
+	// Whether the assignment has been triggered already.
+	triggered: bool,
+}
+
+impl OurAssignment {
+	pub(crate) fn cert(&self) -> &AssignmentCert {
+		&self.cert
+	}
+
+	pub(crate) fn tranche(&self) -> DelayTranche {
+		self.tranche
+	}
+
+	pub(crate) fn validator_index(&self) -> ValidatorIndex {
+		self.validator_index
+	}
+
+	pub(crate) fn triggered(&self) -> bool {
+		self.triggered
+	}
+
+	pub(crate) fn mark_triggered(&mut self) {
+		self.triggered = true;
+	}
+}
+
+fn relay_vrf_modulo_transcript(
+	relay_vrf_story: RelayVRFStory,
+	sample: u32,
+) -> Transcript {
+	// combine the relay VRF story with a sample number.
+	let mut t = Transcript::new(approval_types::RELAY_VRF_MODULO_CONTEXT);
+	t.append_message(b"RC-VRF", &relay_vrf_story.0);
+	sample.using_encoded(|s| t.append_message(b"sample", s));
+
+	t
+}
+
+fn relay_vrf_modulo_core(
+	vrf_in_out: &VRFInOut,
+	n_cores: u32,
+) -> CoreIndex {
+	let bytes: [u8; 4] = vrf_in_out.make_bytes(approval_types::CORE_RANDOMNESS_CONTEXT);
+
+	// interpret as little-endian u32.
+	let random_core = LE::read_u32(&bytes) % n_cores;
+	CoreIndex(random_core)
+}
+
+fn relay_vrf_delay_transcript(
+	relay_vrf_story: RelayVRFStory,
+	core_index: CoreIndex,
+) -> Transcript {
+	let mut t = Transcript::new(approval_types::RELAY_VRF_DELAY_CONTEXT);
+	t.append_message(b"RC-VRF", &relay_vrf_story.0);
+	core_index.0.using_encoded(|s| t.append_message(b"core", s));
+	t
+}
+
+fn relay_vrf_delay_tranche(
+	vrf_in_out: &VRFInOut,
+	num_delay_tranches: u32,
+	zeroth_delay_tranche_width: u32,
+) -> DelayTranche {
+	let bytes: [u8; 4] = vrf_in_out.make_bytes(approval_types::TRANCHE_RANDOMNESS_CONTEXT);
+
+	// interpret as little-endian u32 and reduce by the number of tranches.
+	let wide_tranche = LE::read_u32(&bytes) % (num_delay_tranches + zeroth_delay_tranche_width);
+
+	// Consolidate early results to tranche zero so tranche zero is extra wide.
+	wide_tranche.saturating_sub(zeroth_delay_tranche_width)
+}
```

Comment:
We need `zeroth_delay_tranche_width` in `RelayEquivocation` but here we'll just add more samples to the modulo case since that gives a more uniform distribution.

Reply:
@burdges We could set `zeroth_delay_tranche_width` to 0 in the configuraiton for a little extra flexibility.

Reply:
I'm kinda tempted to say all delays start at tranche 1.  We could add 1 to wide_tranche so that subtracting `zeroth_delay_tranche_width=0` leaves everything starting at 1, while `zeroth_delay_tranche_width=1` starts everything at 0, and `zeroth_delay_tranche_width>1` saturates.  I vaguely remember some hiccup here, but maybe it goes away with your separation of this code path from `RelayEquivocation`.. or maybe I'm miss-remembering.